### ISSUE TITLE
MacOS crosscompile not copying required .dylib & trying to copy wrong arch .dylib

### ISSUE THREAD

**Issue by laggykiller** [Nuitka/Nuitka#2469]
I am trying to crosscompile on x86_64 MacOS, with target architecture of arm64.

Some pip packages contain .dylib (`av` and `PIL` in my case), which nuitka is not copying required `.dylib` (For `PIL`), or trying to copy `.dylib` with wrong architecture (For `av`).

TLDR at bottom

## Common information

- Nuitka version, full Python version, flavor, OS, etc. as output by *this exact* command.

  > python -m nuitka --version

```
1.8.3
Commercial: None
Python: 3.11.5 (v3.11.5:cce6ba91b3, Aug 24 2023, 10:50:31) [Clang 13.0.0 (clang-1300.0.29.30)]
Flavor: CPython Official
Executable: /Users/laggykiller/Downloads/av-test/venv/bin/python
OS: Darwin
Arch: x86_64
Version C compiler: /usr/bin/clang (clang).
```

- How did you install Nuitka and Python

Nuitka by pip

Python from python.org .pkg installer

- Below is `compile.py`, which I will refer to later

<details>
  <summary>compile.py</summary>

```python
#!/usr/bin/env python3
import os
import shutil
import subprocess

def osx_run_in_venv(cmd, get_stdout=False):
    if os.path.isfile('/bin/zsh'):
        sh_cmd = ['/bin/zsh', '-c']
    else:
        sh_cmd = ['/bin/bash', '-c']
    venv_cmd = 'source venv/bin/activate && '

    if get_stdout:
        return subprocess.run(sh_cmd + [venv_cmd + cmd], stdout=subprocess.PIPE).stdout.decode()
    else:
        return subprocess.run(sh_cmd + [venv_cmd + cmd])

def search_wheel_in_dir(package: str, dir: str):
    for i in os.listdir(dir):
        if i.startswith(package):
            return i

def copy_if_universal(wheel_name: str, in_dir: str, out_dir: str):
    if wheel_name.endswith('universal2.whl') or wheel_name.endswith('any.whl'):
        src_path = os.path.join(in_dir, wheel_name)
        dst_path = os.path.join(
            out_dir, 
            wheel_name
            .replace('x86_64', 'universal2')
            .replace('arm64', 'universal2')
            )

        shutil.copy(src_path, dst_path)
        return True
    else:
        return False

def create_universal_wheels(in_dir1, in_dir2, out_dir):
    for wheel_name_1 in os.listdir(in_dir1):
        package = wheel_name_1.split('-')[0]
        wheel_name_2 = search_wheel_in_dir(package, in_dir2)
        if copy_if_universal(wheel_name_1, in_dir1, out_dir):
            continue
        if copy_if_universal(wheel_name_2, in_dir2, out_dir):
            continue
        
        wheel_path_1 = os.path.join(in_dir1, wheel_name_1)
        wheel_path_2 = os.path.join(in_dir2, wheel_name_2)
        subprocess.run(['delocate-fuse', wheel_path_1, wheel_path_2, '-w', out_dir])
        print(f'Created universal wheel {wheel_path_1} {wheel_path_2}')

    for wheel_name in os.listdir(out_dir):
        wheel_name_new = wheel_name.replace('x86_64', 'universal2').replace('arm64', 'universal2')

        src_path = os.path.join(out_dir, wheel_name)
        dst_path = os.path.join(out_dir, wheel_name_new)

        os.rename(src_path, dst_path)
        print(f'Renamed universal wheel {dst_path}')

subprocess.run(['python3', '-m', 'venv', 'venv'])

shutil.rmtree('wheel_arm', ignore_errors=True)
shutil.rmtree('wheel_x64', ignore_errors=True)
shutil.rmtree('wheel_universal2', ignore_errors=True)

os.mkdir('wheel_arm')
os.mkdir('wheel_x64')
os.mkdir('wheel_universal2')

osx_run_in_venv('python -m pip install nuitka==1.8.3 pyinstaller==6.0.0 delocate==0.10.4')
osx_run_in_venv('python -m pip download -r requirements.txt --platform macosx_11_0_arm64 --only-binary=:all: -d wheel_arm')
osx_run_in_venv('python -m pip download -r requirements.txt --platform macosx_11_0_x86_64 --only-binary=:all: -d wheel_x64')

create_universal_wheels('./wheel_arm', './wheel_x64', 'wheel_universal2')
osx_run_in_venv(f'python -m pip install ./wheel_universal2/*')

# osx_run_in_venv('clang -arch arm64 -shared -undefined dynamic_lookup -o ./stub-arm64.dylib ./stub.c')
# osx_run_in_venv('clang -arch x86_64 -shared -undefined dynamic_lookup -o ./stub-x86_64.dylib ./stub.c')

# stub_x64 = './stub-x86_64.dylib'
# stub_arm = './stub-arm64.dylib'

# for dirpath, dirnames, filenames in os.walk('venv/lib'):
#     for f in filenames:
#         if os.path.splitext(f)[1] == '.dylib':
#             f_path = os.path.join(dirpath, f)
#             out = osx_run_in_venv(f'lipo -info {f_path}'.split(' '), get_stdout=True)
#             if f_path.endswith('-x86_64.dylib') or f_path.endswith('-arm64.dylib'):
#                 continue
#             if 'x86_64' in out and 'arm64' in out:
#                 continue
#             if 'x86_64' in out:
#                 f_bak = os.path.splitext(f)[0] + '-x86_64.dylib'
#                 f_bak_path = os.path.join(dirpath, f_bak)
#                 os.rename(f_path, f_bak_path)
#                 osx_run_in_venv(f'lipo {f_bak_path} {stub_arm} -create -output {f_path}')
#                 print(f'Created fat library {f}')
#             elif 'arm64' in out:
#                 f_bak = os.path.splitext(f)[0] + '-arm64.dylib'
#                 f_bak_path = os.path.join(dirpath, f_bak)
#                 os.rename(f_path, f_bak_path)
#                 osx_run_in_venv(f'lipo {f_bak_path} {stub_x64} -create -output {f_path}')
#                 print(f'Created fat library {f}')

compile_cmd = 'python -m nuitka --macos-target-arch=arm64 --standalone --follow-imports --macos-create-app-bundle --include-package=av --include-module=av.audio.codeccontext --include-module=av.video.codeccontext test.py'
osx_run_in_venv(compile_cmd)

compile_cmd = 'python -m PyInstaller --target-architecture=arm64 --windowed -n test-pyinstaller test.py'
osx_run_in_venv(compile_cmd)
```
</details>

What `compile.py` does:
1. Create venv environment and use it
2. Download (without install) packages in `requirements.txt`
  - arm64 version in `wheel_arm`
  - x86_64 version in `wheel_x64`
3. Use `delocate` to combine wheels into universal2 wheels and place into `wheel_universal2`
4. Install universal2 wheels in `wheel_universal2`
5. (Commented out section explained in case 2)
6. Create app bundle for `test.py` using nuitka and pyinstaller.

Note that this method works in pyinstaller for creating universal2/arm64 app bundle on x86_64. Actually, this method also works in Nuitka mostly, except for the edge cases below:

## Case 1: `PIL` not copying required  `.dylib`

- The specific PyPI names and versions

requirements.txt
```
Pillow==10.0.1
```

test.py
```
from PIL import Image

img_obj = Image.new('RGBA', size=(1,1))
print(img_obj)
```

Running without compiling (No problem)
```
<PIL.Image.Image image mode=RGBA size=1x1 at 0x1050D8890>
```

Compile with `compile.py`. The command for nuitka would be `python -m nuitka --macos-target-arch=arm64 --standalone --follow-imports --macos-create-app-bundle test.py`

Running on arm64 macOS complains about missing `libjpeg.62.4.0.dylib`:
```
Traceback (most recent call last):
  File "/Users/laggykiller/Downloads/test.app/Contents/MacOS/test.py", line 1, in <module>
  File "/Users/laggykiller/Downloads/test.app/Contents/MacOS/PIL/Image.py", line 82, in <module PIL.Image>
ImportError: dlopen(/Users/laggykiller/Downloads/test.app/Contents/MacOS/PIL/_imaging.so, 0x0002): Library not loaded: @loader_path/.dylibs/libjpeg.62.4.0.dylib
  Referenced from: <ED4A34E4-2022-31D5-929C-88577059BEB1> /Users/laggykiller/Downloads/test.app/Contents/MacOS/PIL/_imaging.so
  Reason: tried: '/Users/laggykiller/Downloads/test.app/Contents/MacOS/PIL/.dylibs/libjpeg.62.4.0.dylib' (no such file), '/usr/local/lib/libjpeg.62.4.0.dylib' (no such file), '/usr/lib/libjpeg.62.4.0.dylib' (no such file, not in dyld cache)
```

Let's look into site-packages directory for PIL. `PIL/.dylibs` contain many `.dylib`. Most of the dylibs exist in both x86_64 and arm64, except `libjpeg.62.4.0.dylib` which exist in arm64 wheel but not x86_64.

For PyInstaller, all the dylib are copied in, producing a working app bundle.

We can repair the app bundle by adding this to the end of `compile.py`. This would copy libjpeg to the app bundle and resign it

```python
site = osx_run_in_venv("python -c 'import site; print(site.getsitepackages()[0])'", get_stdout=True).strip()
pil_dylibs = os.path.join(site, 'PIL/.dylibs')
for i in os.listdir(pil_dylibs):
    if 'libjpeg' in i and not i.endswith('-arm64.dylib') and not i.endswith('-x86_64.dylib'):
        libjpeg_name = i
        libjpeg_path = os.path.join(pil_dylibs, i)
        break
os.makedirs('test.app/Contents/MacOS/PIL/__dot__dylibs', exist_ok=True)
shutil.copy(libjpeg_path, 'test.app/Contents/MacOS')
shutil.copy(libjpeg_path, f'test.app/Contents/MacOS/PIL/__dot__dylibs/{libjpeg_name}')
os.symlink(src='__dot__dylibs', dst='.dylibs', target_is_directory=True, dir_fd=os.open('test.app/Contents/MacOS/PIL', os.O_RDONLY))
osx_run_in_venv(f'codesign --force --deep -s - test.app')
```

Also refer to https://github.com/Nuitka/Nuitka/issues/1568 for creating symlink from `__dot__dylibs` to `.dylibs`

## Case 2: `av` trying to copy `.dylib` with wrong architecture

- The specific PyPI names and versions

requirements.txt
```
av==10.0.0
```

test.py
```
import av

empty_frame = av.VideoFrame()
print(empty_frame)
```

Running without compiling (No problem):
```
<av.VideoFrame #0, pts=None yuv420p 0x0 at 0x1094141c0>
```

Compile with `compile.py`. The command for nuitka would be `python -m nuitka --macos-target-arch=arm64 --standalone --follow-imports --macos-create-app-bundle --include-package=av --include-module=av.audio.codeccontext --include-module=av.video.codeccontext test.py`

Compilation fails with nuitka complaining it cannot find arm64 version `libX11.6.dylib`:
```
FATAL: Error, cannot use file '/Users/laggykiller/Downloads/av-test/venv/lib/python3.11/site-packages/av/.dylibs/libX11.6.dylib' (Mach-O 64-bit dynamically linked shared library x86_64) to build arch 'arm64' result
```

In reality, there is no libX11 on arm64 macOS, and there is no `libX11.6.dylib` in arm64 wheel (Opposite to the case of `PIL`).

For PyInstaller, it again copied in all the dylib, producing a working app bundle.

We can fool Nuitka to not complain about dylibs that do not have target architecture version.

Create `stub.c`:
```
int main() {
   return 0;
}
```

**Uncomment the section in compile.py**. The section that was commented out do the following:
1. Compiles `stub.c` into `stub-arm64.dylib` and `stub-x86_64.dylib`.
2. Search for dylibs that only contain one architecture with `lipo -info`
3. Use `lipo` to combine those dylibs with `stub-xxx.dylib` to create a 'universal' dylib

This can fool Nuitka to include those dylibs and suppress error. This results in a functioning app bundle that can run on arm64 macOS.

## TLDR
- Some packages may contain .dylib
- Some packages have different .dylib for different architecture of wheel
- Nuitka does not copy .dylib for the target architecture even it is inside `site-packages` (See the case of `PIL`)
- When libfoo.dylib is requirement for x86_64 but not for arm64, Nuitka would complain that it cannot find arm64 version of libfoo.dylib when in reality it is not required and non-existent (See the case of `av`).
- **The whole problem seems to originate from Nuitka assuming if libfoo.dylib is requirement on host architecture, then libfoo.dylib is also requirement on target architecture, which is actually not true.**
- PyInstaller would copy all .dylib for a package with no questions asked (and use lipo to trim away unused architecture from universal library), and it just works.
- If we are not following what PyInstaller is doing, we can instead check if the `.dylib` matches target architecture using `lipo -info xxx.dylib`.

(implementating `getDllExportedSymbols` may help solve this issue?  https://github.com/Nuitka/Nuitka/issues/2468#issuecomment-1742036555)

**Comment by kayhayen** [comment]
Nuitka already checks architectures and trims architectures of files with a target arch, if the compiling Python and/or Python packages are universal arch. I think with `arch -c x86_64` I should be able to test what you are doing there with my M2 machine. I do not have access to the older system anymore.

I am surprised to see cross compilation in this direction happening. But it's not unlikely that the other direction has the exact same issues.

But Nuitka does use `otool` etc. without considering target arch, although that tool has `-arch arch_type` and that is maybe all there is to it. I think the idea was then, that these would be fat DLLs, and that's what I saw so far, rather than 2 files. This also affects #2470 in a way, which too has that argument.

**Comment by kayhayen** [comment]
Hm, seems --list-package-dlls doesn't fill in the options argument for the target arch.

**Comment by kayhayen** [comment]
So, delayed the call of tools to when the options object will be working for them, but now I got this:

```
nm -D /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/lxml/_elementpath.cpython-310-darwin.so --arch=arm64
/Library/Developer/CommandLineTools/usr/bin/nm: error: /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/lxml/_elementpath.cpython-310-darwin.so: File format has no dynamic symbol table
```

So, `nm -D` won't work, nor does `objdump -p` on macOS, very bad.

**Comment by kayhayen** [comment]
Ok, but `nm -gU` works on macOS, and only there, my goodness, but with that, this one is solved, and with the arch argument used as well, hopefully for this issue too.

**Comment by laggykiller** [comment]
Good to here about that, but actually using nm is only faster than lief by just a bit

<details>
  <summary>with_nm.py</summary>

```python
import os
import site
import subprocess

LIB_EXT = ('.dylib', '.so', '.dll', '.pyd')

def is_py_module(lib_path):
    result = subprocess.run(['nm', lib_path], stdout=subprocess.PIPE).stdout.decode()
    if 'PyInit_' in result:
        return True
        
    return False

def get_pkg_root(pkg):
    sites = site.getsitepackages()
    user_site = site.getusersitepackages()
    sites.append(user_site)
    for site_package_dir in sites:
        if pkg in os.listdir(site_package_dir):
            pkg_root = os.path.join(site_package_dir, pkg)
            return pkg_root
    
    return None

def get_non_extension_lib(pkg):
    pkg_root = get_pkg_root(pkg)
    libs = []
    for dirpath, dirnames, filenames in os.walk(pkg_root):
        if '__pycache__' in dirpath:
            continue
        for f in filenames:
            if os.path.splitext(f)[1] in LIB_EXT:
                lib_path = os.path.join(dirpath, f)
                if not is_py_module(lib_path):
                    libs.append(lib_path)

    return libs

print(get_non_extension_lib('av'))
```

</details>

<details>
  <summary>with_lief.py</summary>

```python
import os
import site
import lief

LIB_EXT = ('.dylib', '.so', '.dll', '.pyd')

def is_py_module(lib_path):
    lib = lief.parse(lib_path)
    for i in lib.exported_functions:
        if 'PyInit_' in i.name:
            return True
        
    return False

def get_pkg_root(pkg):
    sites = site.getsitepackages()
    user_site = site.getusersitepackages()
    sites.append(user_site)
    for site_package_dir in sites:
        if pkg in os.listdir(site_package_dir):
            pkg_root = os.path.join(site_package_dir, pkg)
            return pkg_root
    
    return None

def get_non_extension_lib(pkg):
    pkg_root = get_pkg_root(pkg)
    libs = []
    for dirpath, dirnames, filenames in os.walk(pkg_root):
        if '__pycache__' in dirpath:
            continue
        for f in filenames:
            if os.path.splitext(f)[1] in LIB_EXT:
                lib_path = os.path.join(dirpath, f)
                if not is_py_module(lib_path):
                    libs.append(lib_path)

    return libs

print(get_non_extension_lib('av'))
```

</details>

On macOS x86_64, checking for non-extension libraries of `av`
with_nm.py: 2.084s
with_lief.py: 3.044s

**Comment by kayhayen** [comment]
Forking external processes being faster is always a good thing. I guess, that is because `parse` will do a lot more than necessary, but the time is not as critical, since we will be able to cache it anyway if it were a problem, so the deciding factory for me is rather that things like `nm` are widely used and extremely correct usually.

That said, using `nm` is only a thing of the future, the fix for this is probably in how `otool` is being used and I think that might make all the difference for the error, since otool probably defaulted on the host architecture, and now Nuitka will default on that, but our otool usage will honor `--macos-target-architecture` if given, or the running Python architecture.

**Comment by kayhayen** [comment]
It does factor in somewhat, scanning DLLs should detect architecture of the binaries as well, i.e. ignore DLLs not really for that architecture. We already have code to detect architectures in at least Scons (looking at Python.exe 32/64 bits), etc. but that is not the focus right now, I will however want to be able to give a pattern for DLLs and have Nuitka sort out the arch at some point.

I clearly do not like this kind of config very much:

```
- module-name: 'webview'
  data-files:
    dirs:
      - 'lib'
  dlls:
    - from_filenames:
        relative_path: 'lib/runtimes/win-x86/native'
        prefixes:
          - 'WebView2Loader'
      when: 'win32 and arch_x86'
    - from_filenames:
        relative_path: 'lib/runtimes/win-x64/native'
        prefixes:
          - 'WebView2Loader'
      when: 'win32 and arch_amd64'
    - from_filenames:
        relative_path: 'lib/runtimes/win-arm64/native'
        prefixes:
          - 'WebView2Loader'
      when: 'win32 and arch_arm64'
    - from_filenames:
        relative_path: 'lib/x86'
        prefixes:
          - 'WebView2Loader'
      when: 'win32 and arch_x86'
    - from_filenames:
        relative_path: 'lib/x64'
        prefixes:
          - 'WebView2Loader'
      when: 'win32 and arch_amd64'
    - from_filenames:
        relative_path: 'lib/arm64'
        prefixes:
          - 'WebView2Loader'
      when: 'win32 and arch_arm64'
    - from_filenames:
        relative_path: 'lib'
        prefixes:
          - 'Microsoft.'
      when: 'win32'
    - from_filenames:
        relative_path: 'lib'
        prefixes:
          - 'WebBrowserInterop.x86'
      when: 'win32 and arch_x86'
    - from_filenames:
        relative_path: 'lib'
        prefixes:
          - 'WebBrowserInterop.x64'
      when: 'win32 and arch_amd64'
```

**Comment by kayhayen** [comment]
In above config, Nuitka checks (Windows) target arch and selects files. It would be much nicer to just specify all the paths, or patterns, or maybe all DLLs for that package, and have it figure out that the ARM64 binary is not a candidate. There will of course have to be exceptions, e.g. for dotnet you need to include PE files still even on Linux, in then we can declare them as data files, because that is truly what they are then

```
- module-name: 'pythonnet'
  # These is a dotnet assembly, not a platform DLL, esp. not on Linux, macOS,
  # etc.
  data-files:
    patterns:
      - 'runtime/Python.Runtime.dll'
```

So, all of that is actually going to be exciting. Regarding the cross compilation of yours, I believe my change on factory ought to make things different for you now. Give it a shot please.

**Comment by laggykiller** [comment]
Tested with factor branch just now, unfortunately `av` is still problematic:
```
FATAL: Error, cannot use file '/Users/mcchau/Downloads/sticker-convert/venv/lib/python3.11/site-packages/av/.dylibs/libX11.6.dylib' (Mach-O 64-bit dynamically linked shared library x86_64) to build arch 'arm64' result
```

**Comment by laggykiller** [comment]
Tested factory version just now, can confirm that Nuitka no longer complain about not finding arm64 version of libX11 (`av`) and libjpeg included without intervention (`PIL`)

**Comment by kayhayen** [comment]
Thanks for the feedback, that's great news.

**Comment by kayhayen** [comment]
This is now on the develop branch and part of the latest pre-release, will be part of the 1.9 release that will eventually follow.

**Comment by kayhayen** [comment]
Part of the 1.9 release I just made.
